File: new/two_sum_II_hash_map_one_go.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        // idea: hashmap one go
        // iterate through each number (our current number)
        // for each number find the complement that adds up to the target
        // if the complement exists in the hashmap, return the complement's and then the current number's index (i + 1)
        // else add the current number to the hashmap (so when we go for numbers down the array, this will
        // potentially show up as a complment for our current number at that time). 
        unordered_map<int, int> val_to_index;
        for (int i = 0; i < numbers.size(); i++) {
            int complement = target - numbers[i];
            if (val_to_index.count(complement)) {
                return {val_to_index[complement], i + 1};
            }
            val_to_index[numbers[i]] = i + 1;
        }
        return {};
    }
    // time: O(n)
    // space: O(n)
};

File: new/python/two_sum_II_hash_map_one_go.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        # idea: using hashmap, one go
        # we are using hashmap of {current number: its' index, ...}
        # iterate through the given array
        # for every current number we are currently iterating through, check if the hashmap contains the complement,
        # if we have the complement, return the complement, and the current number.
        # if not add the current number and its index (i + 1) to the hashmap.
        seen = defaultdict(int)  # space: O(n)
        for i in range(len(numbers)):  # O(n)
            complement = target - numbers[i]
            if seen[complement]:  # O(1) hash look up O(1) on average
                return [seen[complement], i + 1]

            seen[numbers[i]] = i + 1
        return []
        # time: O(n)
        # space: O(n)
